devicetypes/smartthings/samsung-smart-doorlock.src/samsung-smart-doorlock.groovy

```diff
@@ -0,0 +1,402 @@
+        int p = pct * 100
+		result = Math.min(100, p)
+        log.debug "${linkText} battery is ${result.value}%"
+	}
+	return result
+}
+
+
+
+/**
+ * Responsible for handling command responses
+ *
+ * @param description The description to be parsed
+ *
+ * @return result: The list of events to be sent out
+ */
+private def parseCommandResponse(String description) {
+	Map descMap = zigbee.parseDescriptionAsMap(description)
+	def deviceName = device.displayName
+	log.trace "ZigBee DTH - Executing parseCommandResponse() for device ${deviceName}"
+
+	def result = []
+	Map responseMap = [:]
+	def data = descMap.data
+
+	def cmd = descMap.commandInt
+	def clusterInt = descMap.clusterInt
+
+	if (clusterInt == CLUSTER_DOORLOCK && (cmd == DOORLOCK_CMD_LOCK_DOOR || cmd == DOORLOCK_CMD_UNLOCK_DOOR)) {
+		log.trace "ZigBee DTH - Executing DOOR LOCK/UNLOCK SUCCESS for device ${deviceName} with description map:- $descMap"
+        //Read the unlock result here, and then confirms that door whether successfully opened or not.
+        if(Integer.parseInt(data[0], 10) == 0){
+        	responseMap.name = "lock"
+			responseMap.displayed = true
+			responseMap.isStateChange = true
+        	responseMap = [ name: "lock", value: "unlocked", descriptionText: "Successfully unlocked" ]
+        }else if(Integer.parseInt(data[0], 10) == 1){
+        	log.debug "failed to unlcok doorlock"
+        }else{
+        	log.debug "unexpected result from unlock command"
+        }
+        
+	} else if (clusterInt == 0x0000){
+    	log.trace "ZigBee DTH - Reading Doorlock Model Name for display purpose ${data}"
+        //Read the model code here, and then translate hex to alphabet. 
+        int length = Integer.parseInt(data[0], 16)
+        int i = 2;
+        String model = "" + (char) Integer.parseInt(data[1], 16)
+        char tempChar
+        while ((Integer.parseInt(data[i], 16) != 32) && i < (length-1)){ 
+			tempChar = (char) Integer.parseInt(data[i], 16)
+            model = model + tempChar
+            i++
+        }
+        responseMap = [ name: "displayName", value: model, displayed: false]
+    } else if (clusterInt == CLUSTER_DOORLOCK && cmd == DOORLOCK_RESPONSE_OPERATION_EVENT) {
+		log.trace "ZigBee DTH - Executing DOORLOCK_RESPONSE_OPERATION_EVENT for device ${deviceName} with description map:- $descMap"
+		def eventSource = Integer.parseInt(data[0], 16)
+		def eventCode = Integer.parseInt(data[1], 16)
+
+		responseMap.name = "lock"
+		responseMap.displayed = true
+		responseMap.isStateChange = true
+
+		def desc = ""
+		def codeName = ""
+
+		if (eventSource == 0) {
+         	desc = "using keypad"
+			responseMap.data = [ method: "keypad" ]
+		} else if (eventSource == 1) {
+			responseMap.data = [ method: "command" ]
+		} else if (eventSource == 2) {
+			desc = "from inside"
+			responseMap.data = [ method: "manual" ]
+		} else if (eventSource == 3) {
+			desc = "using keycard"
+            responseMap.data = [ method: "rfid" ]
```

Comment:
The methods from rfid on down need to be added to the lock capability

Reply:
Can you please give me little more detail regards to your comment above? Thank you.

Reply:
The allowed methods are defined in the Lock capability: https://smartthings.developer.samsung.com/develop/api-ref/capabilities.html#Lock. Any method used by the DTH needs to be added to that capability.

Reply:
@dkirker , could you open a PR to add these additional methods to the capability?

Reply:
Actually NM, I can take care of it